[deploy] torch::deploy API

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51754 [deploy] torch::deploy API**

This API allows you to manage multiple python interpreters in a single
process to deploy PyTorch models packaged with torch.package.

torch/csrc/deploy/deploy.h contains the API definition
torch/csrc/deploy/test_deploy.cpp has some examples.

Notes:
* mutex is added to PyTorchStreamReader to make it safe to use from multiple threads at once.
* USE_DEPLOY is only true for the special libtorch_deployinterpreter.so library, when enabled
  we use a hash table to maintain PyObject <> at::Tensor mappping rather than the internal pointer
  in Tensor since >1 interpreter may have a reference to the tensor.
* serialization.py has some additional functions for creating pickle objects
  but keeping storages in memory for use transfering tensors between interpreters

Differential Revision: [D26329468](https://our.internmc.facebook.com/intern/diff/D26329468)

zdevito committed:
[deploy] torch::deploy API

This API allows you to manage multiple python interpreters in a single
process to deploy PyTorch models packaged with torch.package.

torch/csrc/deploy/deploy.h contains the API definition
torch/csrc/deploy/test_deploy.cpp has some examples.

Notes:
* mutex is added to PyTorchStreamReader to make it safe to use from multiple threads at once.
* USE_DEPLOY is only true for the special libtorch_deployinterpreter.so library, when enabled
  we use a hash table to maintain PyObject <> at::Tensor mappping rather than the internal pointer
  in Tensor since >1 interpreter may have a reference to the tensor.
* serialization.py has some additional functions for creating pickle objects
  but keeping storages in memory for use transfering tensors between interpreters

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -53,6 +53,32 @@ static const char* VOLATILE_WARNING =
+// there are muliple copies of the python interpreter that
+// can shared Tensors, so rather than use their internal pointer
+// to a PyObject use a library-local map.
+static std::unordered_map<void*, PyObject*> impl_to_pyobj;
```
I would have a couple questions about this :) 
- Why is this only done for Tensor? The same thing is true for all the python bindings we have no (for autograd, thinking about Node in particular)?
- Why is this not implemented in `variable.cpp` in the implementation of torch::autograd::impl::pyobj? That would feel better to have all the definition of these functions at the same place.

zdevito replied:
1. AFAIK, other python bindings do not use internal function pointers. Does Node do this? I know that storage doesn't, as well as all the stuff using Pybind. If so, then we need similar code for those. Autograd is not supported in this patch for other subtler reasons inside the engine code itself that assumes a single python, so fixes to Node can come later.
2. The code needs to be defined in `libtorch_python.so`, the python bindings, not in `libtorch_cpu.so`. There is only one copy of `libtorch_cpu.so` so putting a hash table in there would not give us independent tables per interpreter. Technically, we can move everything into `python_variable.cpp` if that is more clear, but then it further exposes the internal details of TensorImpl.

albanD replied:
Thanks for the clarifications!
For future reference, the Node does the same thing here: https://github.com/pytorch/pytorch/blob/fb07aca7b05edfb9d41bca950425e70d0b29bcbf/torch/csrc/autograd/function.h#L294-L303

zdevito replied:
I haven't noticed in any meaningful tests, including timing JIT doing nothing against Python doing nothing, where you would expect this to show up. However, I am not confident in my tests here to say that we should do this universally. Using pybind11 requires buying into other infrastructure (e.g. std::shared_ptr containers for the objects), so it is not ideal here.

wconstab commented on torch/csrc/deploy/deploy.cpp:
```diff
@@ -0,0 +1,117 @@
+    std::ifstream src(interp_path, std::ios::binary);
+    TORCH_CHECK(
+        src.good(),
+        "Could not load libtorch_deployinterpreter.so, was PyTorch build with USE_DEPLOY=ON?");
```
typo 'built'

wconstab commented on torch/csrc/deploy/deploy.cpp:
```diff
@@ -0,0 +1,117 @@
+  Dl_info info;
+  dladdr((void*)find_interpreter, &info);
+  char buf[PATH_MAX];
+  TORCH_INTERNAL_ASSERT(!*stpncpy(buf, info.dli_fname, PATH_MAX));
```
what an arcane expression- i think it means:  assert( not (pointing to (a null character) ) )  
where stpncpy returns a pointer to a null terminator if it succeeds?  what does it return a pointer to otherwise, a null pointer?

zdevito replied:
That expression is my passive aggressive rant against C's string library where they can't manage to ensure size-limited string copy is null terminated unless you use this weird incantation. In the event the path is too long, the string is not null terminated ad stpncpy will return a valid pointer to a non-null character. Ill make the expression more clear when I land this.

wconstab commented on torch/csrc/deploy/deploy.cpp:
```diff
@@ -0,0 +1,117 @@
+}
+
+#ifndef _WIN32
+static std::string find_interpreter() {
```
was this WAR intended for use only on OSS?   I guess it wouldn't work on buck but not sure.

zdevito replied:
Probably won't work for buck, but it is pretty stable for finding the library in OSS since it just requires that it sits next to libtorch_cpu.so

zdevito committed:
Update on "[deploy] torch::deploy API"


This API allows you to manage multiple python interpreters in a single
process to deploy PyTorch models packaged with torch.package.

torch/csrc/deploy/deploy.h contains the API definition
torch/csrc/deploy/test_deploy.cpp has some examples.

Notes:
* mutex is added to PyTorchStreamReader to make it safe to use from multiple threads at once.
* USE_DEPLOY is only true for the special libtorch_deployinterpreter.so library, when enabled
  we use a hash table to maintain PyObject <> at::Tensor mappping rather than the internal pointer
  in Tensor since >1 interpreter may have a reference to the tensor.
* serialization.py has some additional functions for creating pickle objects
  but keeping storages in memory for use transfering tensors between interpreters

[ghstack-poisoned]

zdevito committed:
Update on "[deploy] torch::deploy API"


This API allows you to manage multiple python interpreters in a single
process to deploy PyTorch models packaged with torch.package.

torch/csrc/deploy/deploy.h contains the API definition
torch/csrc/deploy/test_deploy.cpp has some examples.

Notes:
* mutex is added to PyTorchStreamReader to make it safe to use from multiple threads at once.
* USE_DEPLOY is only true for the special libtorch_deployinterpreter.so library, when enabled
  we use a hash table to maintain PyObject <> at::Tensor mappping rather than the internal pointer
  in Tensor since >1 interpreter may have a reference to the tensor.
* serialization.py has some additional functions for creating pickle objects
  but keeping storages in memory for use transfering tensors between interpreters

[ghstack-poisoned]

zdevito committed:
Update on "[deploy] torch::deploy API"


This API allows you to manage multiple python interpreters in a single
process to deploy PyTorch models packaged with torch.package.

torch/csrc/deploy/deploy.h contains the API definition
torch/csrc/deploy/test_deploy.cpp has some examples.

Notes:
* mutex is added to PyTorchStreamReader to make it safe to use from multiple threads at once.
* USE_DEPLOY is only true for the special libtorch_deployinterpreter.so library, when enabled
  we use a hash table to maintain PyObject <> at::Tensor mappping rather than the internal pointer
  in Tensor since >1 interpreter may have a reference to the tensor.
* serialization.py has some additional functions for creating pickle objects
  but keeping storages in memory for use transfering tensors between interpreters

[ghstack-poisoned]

zdevito committed:
Update on "[deploy] torch::deploy API"


This API allows you to manage multiple python interpreters in a single
process to deploy PyTorch models packaged with torch.package.

torch/csrc/deploy/deploy.h contains the API definition
torch/csrc/deploy/test_deploy.cpp has some examples.

Notes:
* mutex is added to PyTorchStreamReader to make it safe to use from multiple threads at once.
* USE_DEPLOY is only true for the special libtorch_deployinterpreter.so library, when enabled
  we use a hash table to maintain PyObject <> at::Tensor mappping rather than the internal pointer
  in Tensor since >1 interpreter may have a reference to the tensor.
* serialization.py has some additional functions for creating pickle objects
  but keeping storages in memory for use transfering tensors between interpreters

[ghstack-poisoned]

zdevito committed:
Update on "[deploy] torch::deploy API"


This API allows you to manage multiple python interpreters in a single
process to deploy PyTorch models packaged with torch.package.

torch/csrc/deploy/deploy.h contains the API definition
torch/csrc/deploy/test_deploy.cpp has some examples.

Notes:
* mutex is added to PyTorchStreamReader to make it safe to use from multiple threads at once.
* USE_DEPLOY is only true for the special libtorch_deployinterpreter.so library, when enabled
  we use a hash table to maintain PyObject <> at::Tensor mappping rather than the internal pointer
  in Tensor since >1 interpreter may have a reference to the tensor.
* serialization.py has some additional functions for creating pickle objects
  but keeping storages in memory for use transfering tensors between interpreters

[ghstack-poisoned]

zdevito committed:
Update on "[deploy] torch::deploy API"


This API allows you to manage multiple python interpreters in a single
process to deploy PyTorch models packaged with torch.package.

torch/csrc/deploy/deploy.h contains the API definition
torch/csrc/deploy/test_deploy.cpp has some examples.

Notes:
* mutex is added to PyTorchStreamReader to make it safe to use from multiple threads at once.
* USE_DEPLOY is only true for the special libtorch_deployinterpreter.so library, when enabled
  we use a hash table to maintain PyObject <> at::Tensor mappping rather than the internal pointer
  in Tensor since >1 interpreter may have a reference to the tensor.
* serialization.py has some additional functions for creating pickle objects
  but keeping storages in memory for use transfering tensors between interpreters

[ghstack-poisoned]

zdevito committed:
Update on "[deploy] torch::deploy API"


This API allows you to manage multiple python interpreters in a single
process to deploy PyTorch models packaged with torch.package.

torch/csrc/deploy/deploy.h contains the API definition
torch/csrc/deploy/test_deploy.cpp has some examples.

Notes:
* mutex is added to PyTorchStreamReader to make it safe to use from multiple threads at once.
* USE_DEPLOY is only true for the special libtorch_deployinterpreter.so library, when enabled
  we use a hash table to maintain PyObject <> at::Tensor mappping rather than the internal pointer
  in Tensor since >1 interpreter may have a reference to the tensor.
* serialization.py has some additional functions for creating pickle objects
  but keeping storages in memory for use transfering tensors between interpreters

[ghstack-poisoned]

zdevito committed:
Update on "[deploy] torch::deploy API"


This API allows you to manage multiple python interpreters in a single
process to deploy PyTorch models packaged with torch.package.

torch/csrc/deploy/deploy.h contains the API definition
torch/csrc/deploy/test_deploy.cpp has some examples.

Notes:
* mutex is added to PyTorchStreamReader to make it safe to use from multiple threads at once.
* USE_DEPLOY is only true for the special libtorch_deployinterpreter.so library, when enabled
  we use a hash table to maintain PyObject <> at::Tensor mappping rather than the internal pointer
  in Tensor since >1 interpreter may have a reference to the tensor.
* serialization.py has some additional functions for creating pickle objects
  but keeping storages in memory for use transfering tensors between interpreters

[ghstack-poisoned]

zdevito committed:
Update on "[deploy] torch::deploy API"


This API allows you to manage multiple python interpreters in a single
process to deploy PyTorch models packaged with torch.package.

torch/csrc/deploy/deploy.h contains the API definition
torch/csrc/deploy/test_deploy.cpp has some examples.

Notes:
* mutex is added to PyTorchStreamReader to make it safe to use from multiple threads at once.
* USE_DEPLOY is only true for the special libtorch_deployinterpreter.so library, when enabled
  we use a hash table to maintain PyObject <> at::Tensor mappping rather than the internal pointer
  in Tensor since >1 interpreter may have a reference to the tensor.
* serialization.py has some additional functions for creating pickle objects
  but keeping storages in memory for use transfering tensors between interpreters

[ghstack-poisoned]

wconstab commented on torch/csrc/deploy/deploy.cpp:
```diff
@@ -0,0 +1,110 @@
+#include <unistd.h>
+#endif
+
+extern "C" char _binary_libtorch_deployinterpreter_so_start[];
```
perhaps a comment, about why this reference to the 'blank_library'?

wconstab replied:
i figured out part of it so far-

these externs just mean - there are these symbols that point to the begin/end of a binary blob in memory

there is a 'blank_library' file, which i thought was more complicated than it was - i think it's just what it purports to be- a blank library.  it's purpose is so that these symbols are defined (but point to the same address, a bin blob of len 0) in the case of not building torch deploy.

I assume there is some other code that in the torch deploy case actually populates these symbols with the result of the torchdeploy.o object that is created with the  cmake custom command that does 
OUTPUT libtorch_deployinterpreter.o

zdevito replied:
I've switched to building libtorch_deploy.so separately from libtorch_cpu.so for other reasons (adds huge size to the other one, _cpu has lots of build rules associated with it). This means we don't need blank library or windows ifdefs for deploy since deploy.cpp is only built conditionally. Added a comment for `_binary_libtorch_deployinterpreter_so_start`. It is built with the `ld -r -b binary`command.

wconstab commented on torch/csrc/deploy/deploy.cpp:
```diff
@@ -0,0 +1,110 @@
+        "Intepreter library libtorch_deployinterpreter.so was not included, was PyTorch built with USE_DEPLOY=ON?");
+    std::ofstream dst(library_name, std::ios::binary);
+    dst.write(
+        _binary_libtorch_deployinterpreter_so_start,
```
ah, that's a nice way to do it

wconstab commented on torch/csrc/deploy/deploy.cpp:
```diff
@@ -0,0 +1,110 @@
+  if (handle_) {
+    // ensure python uninitialization runs before we dlclose the library
+    pImpl_.reset();
+    // it segfaults its face off trying to unload, but it's not clear
```
is this still true even? might be worth tracking down a little further in case we're masking something?

zdevito replied:
no, deleted

wconstab commented on caffe2/CMakeLists.txt:
```diff
@@ -1754,3 +1754,8 @@ if(BUILD_PYTHON)
 endif()
+
+if(NOT USE_DEPLOY)
+add_library(torch_deployinterpreter_symbols STATIC ../torch/csrc/deploy/blank_library.cpp)
```
ahh this makes sense now.  still, better put a comment here too explaining why

zdevito committed:
Update on "[deploy] torch::deploy API"


This API allows you to manage multiple python interpreters in a single
process to deploy PyTorch models packaged with torch.package.

torch/csrc/deploy/deploy.h contains the API definition
torch/csrc/deploy/test_deploy.cpp has some examples.

Notes:
* mutex is added to PyTorchStreamReader to make it safe to use from multiple threads at once.
* USE_DEPLOY is only true for the special libtorch_deployinterpreter.so library, when enabled
  we use a hash table to maintain PyObject <> at::Tensor mappping rather than the internal pointer
  in Tensor since >1 interpreter may have a reference to the tensor.
* serialization.py has some additional functions for creating pickle objects
  but keeping storages in memory for use transfering tensors between interpreters

Differential Revision: [D26329468](https://our.internmc.facebook.com/intern/diff/D26329468)

[ghstack-poisoned]

dzhulgakov commented on torch/csrc/deploy/interpreter/interpreter_impl.cpp:
```diff
@@ -0,0 +1,455 @@
+using namespace py::literals;
+
+// TODO this should come from cmake
+#define DEBUG 1
```
do you just want to have debug mode? in this case `NDEBUG` (for Release) is defined correctly (it's actually a compiler de-facto standard while `DEBUG` is not)

dzhulgakov commented on torch/csrc/deploy/interpreter/interpreter_impl.cpp:
```diff
@@ -0,0 +1,455 @@
+  }                       \
+  assert(NULL != obj);
+#elif (DEBUG == 0)
+#define PYOBJ_ASSERT(obj) assert(NULL != obj);
```
assert is a noop in release mode anyway

dzhulgakov commented on torch/csrc/deploy/interpreter/interpreter_impl.cpp:
```diff
@@ -0,0 +1,455 @@
+    PyErr_Print();        \
+  }                       \
+  assert(NULL != obj);
+#elif (DEBUG == 0)
```
why not just `else`?

dzhulgakov commented on torch/csrc/deploy/interpreter/interpreter_impl.cpp:
```diff
@@ -0,0 +1,455 @@
+using torch::PickledObject;
+using torch::PythonObject;
+
+struct ScopedAcquire {
```
can it by pybind11's helper instead?

dzhulgakov commented on torch/csrc/deploy/interpreter/interpreter_impl.cpp:
```diff
@@ -0,0 +1,455 @@
+    // thread grabs the GIL to do non-initialization tasks, then it might start
+    // initializing (GIL -> init_lock). To avoid this, releasethe GIL before
+    // trying to get the init_lock and then reacquire it afterward.
+    PyEval_SaveThread();
```
it can be pybind's helper as well

dzhulgakov commented on torch/csrc/deploy/interpreter/interpreter_impl.cpp:
```diff
@@ -0,0 +1,455 @@
+struct ConcreteInterpreterImpl : public torch::InterpreterImpl {
+  ConcreteInterpreterImpl() {
+    // some dependency in mkl requires this...
+    void* result = dlopen("libz.so", RTLD_GLOBAL | RTLD_LAZY);
```
is it a recent thing in OSS build or some earlier workaround for fbcode. There was a fix recently for something libz related (https://fburl.com/bibq1q9z)

dzhulgakov commented on torch/csrc/deploy/interpreter/interpreter_impl.h:
```diff
@@ -1,26 +1,98 @@
+  std::shared_ptr<caffe2::serialize::PyTorchStreamReader> container_file_;
+};
+
+struct PythonObject {
```
some comment on design of PythonObject would be helpful, i.e. why do we need this wrapping

Commit:
Update on "[deploy] torch::deploy API"


This API allows you to manage multiple python interpreters in a single
process to deploy PyTorch models packaged with torch.package.

torch/csrc/deploy/deploy.h contains the API definition
torch/csrc/deploy/test_deploy.cpp has some examples.

Notes:
* mutex is added to PyTorchStreamReader to make it safe to use from multiple threads at once.
* USE_DEPLOY is only true for the special libtorch_deployinterpreter.so library, when enabled
  we use a hash table to maintain PyObject <> at::Tensor mappping rather than the internal pointer
  in Tensor since >1 interpreter may have a reference to the tensor.
* serialization.py has some additional functions for creating pickle objects
  but keeping storages in memory for use transfering tensors between interpreters

Differential Revision: [D26329468](https://our.internmc.facebook.com/intern/diff/D26329468)

[ghstack-poisoned]

zdevito commented on torch/csrc/deploy/deploy.h:
```diff
@@ -0,0 +1,235 @@
+
+struct Package;
+
+struct TORCH_API LoadBalancer {
```
This is what I ended up with after timing scaling of an effectively no-op model. With 64 threads, anything that tries to take locks doesn't scale. Overkill for not very small models, but this logic is pretty simply anyway.

dzhulgakov replied:
I wonder whether in real setup we'd just have one thread per interpreter so then implementation would be to have an efficient queue of requests and each thread trying to pull from it (with something like folly::MPMCQueue making it very efficient). Or do you think overheads per-interpreter are high enough (mostly memory?) to justify sharing interpreters across threads? Even in that case, we can still have just fixed # of threads per interpreter.

zdevito replied:
The intent is for this to be very fast when n_hardware_threads == n_interpreters, but for it to still work correctly in other conditions. This design means that clients of the library don't have to worry about pinning interpreters to threads, or even really realizing that multiple interpreters are involved in the first place. 

We need to avoid having n_interpreters == n_software_threads. Each one of these interpreters takes ~1--2 seconds to startup because its dlopning our python bindings and `import torch`ing. At startup we want to just create enough interpreters to cover our cpu cores and then never have to create another interpreter.

swolchok commented on torch/csrc/deploy/deploy.h:
```diff
@@ -0,0 +1,235 @@
+  }
+#endif
+ private:
+  uint64_t* locks_;
```
this could be a std::unique_ptr<uint64_t[]>

swolchok commented on torch/csrc/deploy/deploy.h:
```diff
@@ -0,0 +1,235 @@
+    memset(locks_, 0, 8 * n_ * sizeof(uint64_t));
+  }
+  void setResourceLimit(size_t n) {
+    n_ = n;
```
no bounds check, even in debug mode? we need to have n <= n_ to guarantee that we avoid an out-of-bounds access, right?

zdevito replied:
The only user of this API is checking this, but I added an assert

swolchok commented on torch/csrc/deploy/deploy.h:
```diff
@@ -0,0 +1,235 @@
+struct Package;
+
+struct TORCH_API LoadBalancer {
+  LoadBalancer(size_t n) : locks_(new uint64_t[8 * n]), n_(n) {
```
seems like it could be std::unique_ptr<uint64_t[]>

swolchok commented on torch/csrc/deploy/deploy.h:
```diff
@@ -0,0 +1,235 @@
+      if (last >= n_) {
+        last = 0;
+      }
+      uint64_t prev =
+          __atomic_fetch_add(&locks_[8 * last], 1ULL, __ATOMIC_SEQ_CST);
+      if (prev == 0) {
+        // fast path, we found an interpreter with no users
+        return last;
+      } else {
+        // slow path, we don't want to use this interpreter because it is being
+        // used by someone else.
+        __atomic_fetch_sub(&locks_[8 * last], 1ULL, __ATOMIC_SEQ_CST);
+      }
```
shouldn't this just be a strong CAS where it expects 0 and wants to write 1? (and thus it's not clear to me that this logic is pretty simple...)

zdevito replied:
The counts in `locks_` are just an estimate of how many threads are using each interpreter, they don't have to be strictly accurate in the short time. Ill rename `locks_` because it isn't an accurate name. 

The reason for doing it this way is so that the fast path doesn't have any retry logic -- it just does a single atomic add and can move on to execution. 

While this balancer only really matters for small models, I think this is still important to include because we are selling this as "GIL-free" python election. The very first thing someone might do to evaluate whether this is working will be to run 64 threads, each doing very little work, and checking that they scale. I measured a number of approaches and any that used a single lock or contended atomics scaled for really small programs.

swolchok replied:
Here is what I'm proposing: https://godbolt.org/z/ao6Yod

There isn't retry logic in my proposal either, it's just a lock cmpxchg instead of a lock xadd, and you don't have to do a lock xsub if the interpreter wasn't free.

swolchok replied:
(It's entirely possible I've missed something or that this is actually a bad idea on x86_64 -- I'm not an expert on the hardware primitives, I just saw two atomic operations instead of one.)

zdevito replied:
My thought was that some architectures implement an atomic add faster than a CAS (cuda), but I don't think x86 is one of them. So using a CAS here would be simpler and I can put that in.

wconstab commented on torch/_deploy.py:
```diff
@@ -0,0 +1,74 @@
+from typing import Callable
+from types import ModuleType
+
+def _save_storages(importer, obj):
```
is the idea here that you could either use torch.package to do the save, or not use it?  I don't really understand torch.package especially what it means to create a custom import pickler without using torch.package's imporer

zdevito replied:
a custom importer will resolve code objects from within the package. The importer here gets set if the MovableObject we are working with was produced from a package.

wconstab commented on torch/csrc/deploy/CMakeLists.txt:
```diff
@@ -1,3 +1,27 @@
+
+add_custom_command(
+  OUTPUT libtorch_deployinterpreter.o
+  COMMAND cp $<TARGET_FILE:torch_deployinterpreter> .
```
is this cp and the following rm necessary? or can you just pass this $<> to ld?

zdevito replied:
the name of the symbol is determined by the filename :( 

I can't find a flag to set otherwise, so I have to do this silly copy.

wconstab commented on torch/csrc/deploy/example/benchmark.cpp:
```diff
@@ -0,0 +1,320 @@
+
+      eg = I.global("builtins", "tuple")(
+                I.self.attr("load_pickle")({"model", "example.pkl"}))
+               .toIValue()
```
why toIValue->toTuple? 

is this to cause a MoveableObject to be created when before it was just a PythonObject?

zdevito replied:
I am just extracting the arguments into the form that a function call will take later. Don't want to use MovableObject here because they you need ensure the `eg` object gets placed on the same interpreter as the model (which I don't think I have a public API for, but should probably be added anyway).

wconstab commented on torch/csrc/deploy/test_deploy.cpp:
```diff
@@ -0,0 +1,126 @@
+  size_t ninterp = 3;
+  std::vector<at::Tensor> outputs;
+
+  // Futures on model forward
```
why do you call these futures? it looks to me like actual output tensors are put onto the vector, but i didn't run the code.

wconstab replied:
nvm guessing its a copy from the other test; should clean up.

wconstab commented on torch/csrc/deploy/test_deploy.cpp:
```diff
@@ -0,0 +1,126 @@
+  torch::InterpreterManager manager(nthreads);
+
+  torch::Package p = manager.load_package(path("SIMPLE", simple));
+  auto model = p.load_pickle("model", "model.pkl");
```
i'm confused about the package and the pickle. i'll go try to track that down.

wconstab commented on torch/csrc/deploy/test_deploy.cpp:
```diff
@@ -0,0 +1,126 @@
+      auto result = model({input}).toTensor();
+      return result;
+    }));
+  }
```
can we also test somehow that we're really parallel?  It seems like this test would still pass if you load-unbalanced to the same interpreter and got GIL locked

wconstab commented on torch/csrc/deploy/test_deploy.cpp:
```diff
@@ -0,0 +1,126 @@
+  }
+}
+
+TEST(TorchpyTest, ThreadedSimpleModel) {
```
not that it would affect actual coverage, but it might be good to include tests of any other expected/popular threading model such as in your paper the OMP for

zdevito committed:
Update on "[deploy] torch::deploy API"


This API allows you to manage multiple python interpreters in a single
process to deploy PyTorch models packaged with torch.package.

torch/csrc/deploy/deploy.h contains the API definition
torch/csrc/deploy/test_deploy.cpp has some examples.

Notes:
* mutex is added to PyTorchStreamReader to make it safe to use from multiple threads at once.
* USE_DEPLOY is only true for the special libtorch_deployinterpreter.so library, when enabled
  we use a hash table to maintain PyObject <> at::Tensor mappping rather than the internal pointer
  in Tensor since >1 interpreter may have a reference to the tensor.
* serialization.py has some additional functions for creating pickle objects
  but keeping storages in memory for use transfering tensors between interpreters

Differential Revision: [D26329468](https://our.internmc.facebook.com/intern/diff/D26329468)

[ghstack-poisoned]

zdevito committed:
Update on "[deploy] torch::deploy API"


This API allows you to manage multiple python interpreters in a single
process to deploy PyTorch models packaged with torch.package.

torch/csrc/deploy/deploy.h contains the API definition
torch/csrc/deploy/test_deploy.cpp has some examples.

Notes:
* mutex is added to PyTorchStreamReader to make it safe to use from multiple threads at once.
* USE_DEPLOY is only true for the special libtorch_deployinterpreter.so library, when enabled
  we use a hash table to maintain PyObject <> at::Tensor mappping rather than the internal pointer
  in Tensor since >1 interpreter may have a reference to the tensor.
* serialization.py has some additional functions for creating pickle objects
  but keeping storages in memory for use transfering tensors between interpreters

Differential Revision: [D26329468](https://our.internmc.facebook.com/intern/diff/D26329468)

[ghstack-poisoned]

wconstab commented on torch/csrc/deploy/example/generate_examples.py:
```diff
@@ -0,0 +1,43 @@
+
+parser = argparse.ArgumentParser(description="Generate Examples")
+parser.add_argument("--install_dir", help="Root directory for all output files")
+parser.add_argument("--fbcode_dir", help="Root directory for all output files")
```
'help' is wrong copypasta - confirm if 'fbcode_dir' is even really needed? it gives access to read from files in the fbcode tree without specifying them as inputs to the target, but doesn't look like it's being used

zdevito replied:
the fbcode custom_rule for python objects drops this argument in, so if it isn't here then it just fails...

wconstab replied:
oh, i think if you add 'strict=true' to the custom_rule then it doesn't.  I should say i know this behavior is true but i forgot the flag name.

wconstab commented on torch/csrc/deploy/interpreter/interpreter_impl.cpp:
```diff
@@ -0,0 +1,463 @@
+class F:
+    def find_spec(self, fullname, path, target=None):
+        if fullname == 'torch._C':
+            return sys.meta_path[1].find_spec('torch._C', None, None)
```
i'm confused about what this is doing.  it looks like 'F' is inserted ahead of default finder, and F just forwards torch._C and mrcn to default.  Is the intent here to filter out finding any other packages?  If so then i think the comment above is misleading: it claims the default importer would skip torch._C.

zdevito replied:
something about the 'target' set for torch._C normally makes it not get found as part of the internal archive. This just blanks out the `target` so that it will be found.

wconstab replied:
ahh i see.

wconstab commented on torch/csrc/deploy/interpreter/interpreter_impl.cpp:
```diff
@@ -0,0 +1,463 @@
+
+import importlib.machinery
+import importlib.util
+spec = importlib.machinery.ModuleSpec('maskrcnn_benchmark', None, is_package=True)  # type: ignore
```
i ended up deleting this from master on torchbench, fyi.  was causing headaches for various people and others were saying it was outdated and we should use the torchvision one.  lmk if you're still relying on it (it should still be there on the v0.1 branch)

zdevito replied:
yeah, that is just from my own stuff, and I will delete it from this patch.

Commit:
Update on "[deploy] torch::deploy API"


This API allows you to manage multiple python interpreters in a single
process to deploy PyTorch models packaged with torch.package.

torch/csrc/deploy/deploy.h contains the API definition
torch/csrc/deploy/test_deploy.cpp has some examples.

Notes:
* mutex is added to PyTorchStreamReader to make it safe to use from multiple threads at once.
* USE_DEPLOY is only true for the special libtorch_deployinterpreter.so library, when enabled
  we use a hash table to maintain PyObject <> at::Tensor mappping rather than the internal pointer
  in Tensor since >1 interpreter may have a reference to the tensor.
* serialization.py has some additional functions for creating pickle objects
  but keeping storages in memory for use transfering tensors between interpreters

Differential Revision: [D26329468](https://our.internmc.facebook.com/intern/diff/D26329468)

[ghstack-poisoned]

wconstab commented on torch/csrc/deploy/interpreter/interpreter_impl.cpp:
```diff
@@ -0,0 +1,463 @@
+    for (size_t i = 0, N = args.size(); i != N; ++i) {
+      m_args[i] = torch::jit::toPyObject(args[i]);
+    }
+    return wrap(call(unwrap(obj), m_args));
```
is wrap/unwrap used to work around some python overhead so the common case uses existing python object references instead of looking up by string again?  i first thought maybe it was like a 'unique_ptr' for py::objects but i don't think that's it.

zdevito replied:
wrap turns a py::object into a PythonObject. PythonObjects are the handles people use outside of interpreter_impl.cpp where pybind11 and python APIs are not allowed because the libraries are hidden.

zdevito commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -53,6 +53,32 @@ static const char* VOLATILE_WARNING =
+
+PyObject* pyobj(const Variable& self) {
+  TORCH_CHECK(self.defined(), "cannot call pyobj() on undefined tensor");
+  return impl_to_pyobj[self.unsafeGetTensorImpl()]; // if it doesn't exist yet
```
probably safer that way to ensure something that checks for null PyObject but doesn't set one isn't leaking an entry into the hash table.

zdevito committed:
Update on "[deploy] torch::deploy API"


This API allows you to manage multiple python interpreters in a single
process to deploy PyTorch models packaged with torch.package.

torch/csrc/deploy/deploy.h contains the API definition
torch/csrc/deploy/test_deploy.cpp has some examples.

Notes:
* mutex is added to PyTorchStreamReader to make it safe to use from multiple threads at once.
* USE_DEPLOY is only true for the special libtorch_deployinterpreter.so library, when enabled
  we use a hash table to maintain PyObject <> at::Tensor mappping rather than the internal pointer
  in Tensor since >1 interpreter may have a reference to the tensor.
* serialization.py has some additional functions for creating pickle objects
  but keeping storages in memory for use transfering tensors between interpreters

Differential Revision: [D26329468](https://our.internmc.facebook.com/intern/diff/D26329468)

[ghstack-poisoned]

zdevito committed:
Update on "[deploy] torch::deploy API"


This API allows you to manage multiple python interpreters in a single
process to deploy PyTorch models packaged with torch.package.

torch/csrc/deploy/deploy.h contains the API definition
torch/csrc/deploy/test_deploy.cpp has some examples.

Notes:
* mutex is added to PyTorchStreamReader to make it safe to use from multiple threads at once.
* USE_DEPLOY is only true for the special libtorch_deployinterpreter.so library, when enabled
  we use a hash table to maintain PyObject <> at::Tensor mappping rather than the internal pointer
  in Tensor since >1 interpreter may have a reference to the tensor.
* serialization.py has some additional functions for creating pickle objects
  but keeping storages in memory for use transfering tensors between interpreters

Differential Revision: [D26329468](https://our.internmc.facebook.com/intern/diff/D26329468)

[ghstack-poisoned]

Commit:
Update on "[deploy] torch::deploy API"


This API allows you to manage multiple python interpreters in a single
process to deploy PyTorch models packaged with torch.package.

torch/csrc/deploy/deploy.h contains the API definition
torch/csrc/deploy/test_deploy.cpp has some examples.

Notes:
* mutex is added to PyTorchStreamReader to make it safe to use from multiple threads at once.
* USE_DEPLOY is only true for the special libtorch_deployinterpreter.so library, when enabled
  we use a hash table to maintain PyObject <> at::Tensor mappping rather than the internal pointer
  in Tensor since >1 interpreter may have a reference to the tensor.
* serialization.py has some additional functions for creating pickle objects
  but keeping storages in memory for use transfering tensors between interpreters

Differential Revision: [D26329468](https://our.internmc.facebook.com/intern/diff/D26329468)

[ghstack-poisoned]

Commit:
Update on "[deploy] torch::deploy API"


This API allows you to manage multiple python interpreters in a single
process to deploy PyTorch models packaged with torch.package.

torch/csrc/deploy/deploy.h contains the API definition
torch/csrc/deploy/test_deploy.cpp has some examples.

Notes:
* mutex is added to PyTorchStreamReader to make it safe to use from multiple threads at once.
* USE_DEPLOY is only true for the special libtorch_deployinterpreter.so library, when enabled
  we use a hash table to maintain PyObject <> at::Tensor mappping rather than the internal pointer
  in Tensor since >1 interpreter may have a reference to the tensor.
* serialization.py has some additional functions for creating pickle objects
  but keeping storages in memory for use transfering tensors between interpreters

Differential Revision: [D26329468](https://our.internmc.facebook.com/intern/diff/D26329468)

[ghstack-poisoned]

zdevito committed:
Update on "[deploy] torch::deploy API"


This API allows you to manage multiple python interpreters in a single
process to deploy PyTorch models packaged with torch.package.

torch/csrc/deploy/deploy.h contains the API definition
torch/csrc/deploy/test_deploy.cpp has some examples.

Notes:
* mutex is added to PyTorchStreamReader to make it safe to use from multiple threads at once.
* USE_DEPLOY is only true for the special libtorch_deployinterpreter.so library, when enabled
  we use a hash table to maintain PyObject <> at::Tensor mappping rather than the internal pointer
  in Tensor since >1 interpreter may have a reference to the tensor.
* serialization.py has some additional functions for creating pickle objects
  but keeping storages in memory for use transfering tensors between interpreters

Differential Revision: [D26329468](https://our.internmc.facebook.com/intern/diff/D26329468)

[ghstack-poisoned]

suo commented on torch/csrc/deploy/interpreter/interpreter_impl.cpp:
```diff
@@ -0,0 +1,469 @@
+};
+
+torch::InterpreterSessionImpl* ConcreteInterpreterImpl::acquire_session() {
+  return new ConcreteInterpreterSessionImpl(this);
```
who frees this?

zdevito committed:
Update on "[deploy] torch::deploy API"


This API allows you to manage multiple python interpreters in a single
process to deploy PyTorch models packaged with torch.package.

torch/csrc/deploy/deploy.h contains the API definition
torch/csrc/deploy/test_deploy.cpp has some examples.

Notes:
* mutex is added to PyTorchStreamReader to make it safe to use from multiple threads at once.
* USE_DEPLOY is only true for the special libtorch_deployinterpreter.so library, when enabled
  we use a hash table to maintain PyObject <> at::Tensor mappping rather than the internal pointer
  in Tensor since >1 interpreter may have a reference to the tensor.
* serialization.py has some additional functions for creating pickle objects
  but keeping storages in memory for use transfering tensors between interpreters

Differential Revision: [D26329468](https://our.internmc.facebook.com/intern/diff/D26329468)

[ghstack-poisoned]
